### ISSUE TITLE
Libnd4j/SameDiff: mmul op: "transpose result" issue

### ISSUE THREAD

**Issue by AlexDBlack** [deeplearning4j/deeplearning4j#5648]
The Mmul op takes a MmulTranspose, which has args to transpose a, b and/or result:
https://github.com/deeplearning4j/deeplearning4j/blob/74bd70162979bace33cf28b160aeb61131e13bcc/nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/blas/params/MMulTranspose.java

However, the libnd4j implementation does NOT support transposing of the result:
https://github.com/deeplearning4j/deeplearning4j/blob/74bd70162979bace33cf28b160aeb61131e13bcc/libnd4j/include/ops/declarable/generic/blas/matmul.cpp

I think we have two options here:
1. Add "transpose result" support to libnd4j matmul op
2. Remove it from Java

We don't seem to need it for (TF) import, as transpose result isn't supported:
https://www.tensorflow.org/api_docs/python/tf/matmul

EDIT: Actually, transpose result *would* be useful for efficiently implementing mmul gradient at the ND4J level (without an extra transpose op). An alternative would be to implement a mmul gradient backprop op at the libnd4j level. But as far as I can see, that would provide negligible performance benefit over an nd4j-based implementation (which, is already done, minus the transposeResult part).

**Comment by shyrma** [comment]
Alex
I have changed matmul op in way it has option "transpose result" now and works correctly in that case (at least in c++ tests :-) )

**Comment by shyrma** [comment]
Hi Alex,
We have added option 1):

**Comment by AlexDBlack** [comment]
Thanks!
I'll leave this issue open until it's properly supported and tested at SameDiff level (presumably some arg changes for mmul op are required)
